spec/queries/keyword_search_entries_query_spec.rb

```diff
@@ -0,0 +1,155 @@
+        keyword = Fabricate(:parsed_keyword) { keyword_search_entries(count: 0) }
+        Fabricate.times(2, :keyword_search_entry, kind: :ads, keyword_id: keyword.id)
+
+        query = described_class.new.with_keyword(keyword.id)
+
+        expect(query.non_ads_count).to(eq(0))
+      end
+    end
+
+    context 'given a keyword with a list of non-ads entries' do
+      it 'returns the number of non-ads entries' do
+        keyword = Fabricate(:parsed_keyword) { keyword_search_entries(count: 0) }
+        Fabricate.times(2, :keyword_search_entry, kind: :non_ads, keyword_id: keyword.id)
+
+        query = described_class.new.with_keyword(keyword.id)
+
+        expect(query.non_ads_count).to(eq(2))
+      end
+    end
+  end
+
+  describe '#top_ads_urls' do
+    context 'given NO top ads search entries' do
+      it 'returns an empty array' do
+        keyword = Fabricate(:parsed_keyword) { keyword_search_entries(count: 0) }
+
+        query = described_class.new.with_keyword(keyword.id)
+
+        expect(query.top_ads_urls).to(be_empty)
+      end
+    end
+
+    context 'given a list of top ads search entries' do
+      # rubocop:disable RSpec/ExampleLength
+      it 'returns all top ads URLs' do
+        keyword = Fabricate(:parsed_keyword) { keyword_search_entries(count: 0) }
+
+        Fabricate(
+          :keyword_search_entry,
+          kind: :ads,
+          position: :top,
+          keyword_id: keyword.id,
+          urls: ['https://google.com', 'https://bing.com']
+        )
+
+        Fabricate(
+          :keyword_search_entry,
+          kind: :ads,
+          position: :top,
+          keyword_id: keyword.id,
+          urls: ['https://www.ruby-lang.org/en/']
+        )
+
+        query = described_class.new.with_keyword(keyword.id)
+
+        expect(query.top_ads_urls).to(contain_exactly('https://google.com', 'https://bing.com', 'https://www.ruby-lang.org/en/'))
+      end
+      # rubocop:enable RSpec/ExampleLength
+    end
+  end
+
+  describe '#non_ads_urls' do
+    context 'given NO non ads search entries' do
+      it 'returns an empty array' do
+        keyword = Fabricate(:parsed_keyword) { keyword_search_entries(count: 0) }
+
+        query = described_class.new.with_keyword(keyword.id)
+
+        expect(query.non_ads_urls).to(be_empty)
+      end
+    end
+
+    context 'given a list of non ads search entries' do
+      # rubocop:disable RSpec/ExampleLength
+      it 'returns all non ads URLs' do
+        keyword = Fabricate(:parsed_keyword) { keyword_search_entries(count: 0) }
+
+        Fabricate(
+          :keyword_search_entry,
+          kind: :non_ads,
+          keyword_id: keyword.id,
+          urls: ['https://google.com', 'https://bing.com']
+        )
+
+        Fabricate(
+          :keyword_search_entry,
+          kind: :non_ads,
+          keyword_id: keyword.id,
+          urls: ['https://www.ruby-lang.org/en/']
+        )
+
+        query = described_class.new.with_keyword(keyword.id)
```

Comment:
Also we seem to test only with the `with_keyword` filter, should we have the test without that? 

```
query = described_class.new

.....
```

Reply:
I just made keyword_id to be a required parameter here 592ac6c25fa3727924b5a9f6d57aa5a7591ec210. It doesn't make sense for our app to query without a keyword. So I think we can omit this test case.